rust/src/linalg/l2.rs

```diff
@@ -149,20 +149,58 @@ mod aarch64 {
         use super::super::l2_scalar;
         use std::arch::aarch64::*;
 
+        // TODO: learn rust macro and refactor to macro instead of manually unroll
+        const UNROLL_FACTOR: usize = 4;
+        const INSTRUCTION_WIDTH: usize = 4;
+
+        const STEP_SIZE: usize = UNROLL_FACTOR * INSTRUCTION_WIDTH;
+
         #[inline]
         pub(crate) fn l2_f32(from: &[f32], to: &[f32]) -> f32 {
             unsafe {
-                let len = from.len() / 4 * 4;
+                let len_aligned_to_unroll = from.len() / STEP_SIZE * STEP_SIZE;
                 let buf = [0.0_f32; 4];
                 let mut sum = vld1q_f32(buf.as_ptr());
-                for i in (0..len).step_by(4) {
+                for i in (0..len_aligned_to_unroll).step_by(STEP_SIZE) {
                     let left = vld1q_f32(from.as_ptr().add(i));
                     let right = vld1q_f32(to.as_ptr().add(i));
                     let sub = vsubq_f32(left, right);
                     sum = vfmaq_f32(sum, sub, sub);
+
+                    let left = vld1q_f32(from.as_ptr().add(i + 4));
+                    let right = vld1q_f32(to.as_ptr().add(i + 4));
+                    let sub = vsubq_f32(left, right);
```

Comment:
wonder are there difference if move the sub/sum to the end. so that you have the memory access instructions together without branch out?

Maybe compiler will do it too? What would be the output from https://rust.godbolt.org/ ?